File: 0806-number-of-lines-to-write-string/0806-number-of-lines-to-write-string.rb
# @param {Integer[]} widths
# @param {String} s
# @return {Integer[]}
def number_of_lines(widths, s)
  costs = ('a'..'z').to_a.zip(widths).to_h
  lines = 1
  width = 0
  
  s.each_char do |chr|
    current = width + costs[chr]
    if current > 100
      lines += 1
      width = costs[chr]
    else
      width = current
    end
  end
  
  [lines, width]
end

File: 0806-number-of-lines-to-write-string/0806-number-of-lines-to-write-string.cpp
class Solution {
public:
    vector<int> numberOfLines(vector<int>& widths, string s) {
        unordered_map<char, int> costs;
        for (int i = 0; i < 26; ++i) costs['a' + i] = widths[i];
        int lines = 1;
        int width = 0;

        for (char chr : s) {
            int current = width + costs[chr];

            if (current > 100) {
                ++lines;
                width = costs[chr];
            } else {
                width = current;
            }
        }

        return {lines, width};
    }
};